samples: google_iot_device: Remove "command line" and replace with config.c

pfalcon opened the pull request:
There is no "command line" in Zephyr RTOS. "Command line" which was
used previously is an artifact of a special-purpose, testing target
"native_posix", not portable to any other Zephyr target.

So, instead provide config.c file, where a user can define all needed
application parameters (which previously were passed on "command
line"). As such a file contain credentials, it should be never put
under version control. This commit includes a template file,
config.c.in, which can be copied by a user to config.c, and needed
values filled in.

Signed-off-by: Paul Sokolovsky <paul.sokolovsky@linaro.org>

pfalcon committed:
samples: google_iot_device: Remove "command line" and replace with config.c

There is no "command line" in Zephyr RTOS. "Command line" which was
used previously is an artifact of a special-purpose, testing target
"native_posix", not portable to any other Zephyr target.

So, instead provide config.c file, where a user can define all needed
application parameters (which previously were passed on "command
line"). As such a file contain credentials, it should be never put
under version control. This commit includes a template file,
config.c.in, which can be copied by a user to config.c, and required
values filled in.

Signed-off-by: Paul Sokolovsky <paul.sokolovsky@linaro.org>

pfalcon commented:
This is based on the discussion in https://github.com/atigyi/zephyr/issues/3#issuecomment-521633236 , and is a prerequisite to getting this sample running on "real" Zephyr device, like `qemu_x86` (`native_posix` is not a real Zephyr device in that regard).

pfalcon commented:
What I'd generally suggest is to completely remove this "command line" handling, i.e. remove code which is so far wrapped in `#if 0` in this PR. Then, apparently rename `commandline.h` to `config.h`.

pfalcon commented:
Rebased on the latest base branch.

@atigyi, This would be on critical path to https://github.com/atigyi/zephyr/issues/2, please share your feedback.

atigyi commented on samples/net/google_iot_device/src/config.c.in:
```diff
@@ -0,0 +1,20 @@
+#include "iotc.h"
+
+/* CHANGE THIS */
+const char* iotc_project_id = "macro-precinct-211108";
```
pls use template strings, something like:
`const char* iotc_project_id = YOUR_GOOGLE_CLOUD_IOT_PROJECT_ID`
Additionally would be good to leave example lines for users. Like this one but with a fictional project name / content. Actually for project name it isn't that useful, but for the `iotc_device_path` I am sure it'll come handy for the first time visitor.

pfalcon replied:
> pls use template strings, something like:

Ack, I just quick-created this template based on the actual config I used ;-). Let me actually take those template strings from: https://github.com/atigyi/zephyr/blob/google_iot_device_sdk_integration/samples/net/google_iot_device/src/main.c#L23 (that comment otherwise needs to be deleted, as there's no longer command line).

atigyi replied:
👍

atigyi commented on samples/net/google_iot_device/src/config.c.in:
```diff
@@ -0,0 +1,20 @@
+"-----END EC PRIVATE KEY-----\n";
+
+const char* iotc_publish_message = "Hello from Zephyr!";
+const char* iotc_private_key_filename = "(none)";
```
`iotc_private_key_filename` won't be required then. This line can be deleted.

pfalcon replied:
> iotc_private_key_filename won't be required then. This line can be deleted.

I put it because it was referenced from somewhere (error message or something). Now that you ok'ed removing unneeded code, I'll make sure to delete it too.

atigyi commented on samples/net/google_iot_device/src/example_utils.c:
```diff
@@ -17,6 +17,7 @@ extern iotc_crypto_key_data_t iotc_connect_private_key_data;
 static iotc_timed_task_handle_t delayed_publish_task =
     IOTC_INVALID_TIMED_TASK_HANDLE;
 
+#if 0
```
These lines can be completely removed from here.

pfalcon replied:
> These lines can be completely removed from here.

Ack!

atigyi commented on samples/net/google_iot_device/src/main.c:
```diff
@@ -42,9 +45,10 @@ void main(void) {
 
-  if (0 != load_ec_private_key_pem_from_posix_fs(ec_private_key_pem,
-                                                 PRIVATE_KEY_BUFFER_SIZE)) {
+  if (0 != load_ec_private_key_pem(ec_private_key_pem,
```
For current state, and sake of simplicity, this function isn't necessary. The `iotc_private_key_pem` could be passed directly to `iotc_connect_private_key_data.crypto_key_union.key_pem.key`.
But if the intention is to show a key-source-ignorant implementation. And/or to prepare for load key from a Secure Element, then this step can stay, doesn't make a huge complexity raise.

pfalcon replied:
> But if the intention is to show a key-source-ignorant implementation. 

Yes, I both wanted to minimize changes, and show that the key loading handled by a separate function, so another implementation can be swapped in.

atigyi replied:
👍

atigyi reviewed (commented):
Looks good, please see some notes. Also please delete the files `commandline.h/.c`. 👍

atigyi commented on samples/net/google_iot_device/src/main.c:
```diff
@@ -26,6 +28,7 @@ void main(void) {
     DeviceID>/state
     */
 
+#if 0
```
Please remove lines in `#if 0 - #endif` blocks.

pfalcon replied:
> Please remove lines in #if 0 - #endif blocks.

Done.

pfalcon commented:
@atigyi : Ok, so I refactored this based on your comments - effectively, went full on with removing command line parsing, and replaced it with C-based config file, down to references in the code comments. As usual, tested with BOARD=native_posix.

Let me know how that looks.

atigyi commented on samples/net/google_iot_device/src/config.c.in:
```diff
@@ -0,0 +1,19 @@
+#include "iotc.h"
```
Please add the copyright + license header to this file.

pfalcon replied:
> Please add the copyright + license header to this file.

Done.

atigyi commented on samples/net/google_iot_device/src/config.c.in:
```diff
@@ -0,0 +1,19 @@
+#include "iotc.h"
```
#include "iotc.h" --> #include <iotc_mqtt.h> ?

pfalcon replied:
> #include "iotc.h" --> #include <iotc_mqtt.h> ?

Hmm, I actually not sure now where `"iotc.h"` comes from, but by the latest changes, it should be `"config.h"`, which declares vars to be defined in config.c[.in]

atigyi commented on samples/net/google_iot_device/src/config.h:
```diff
@@ -0,0 +1,18 @@
+ * Declarations for application config options.
+ */
+
+#include <iotc_mqtt.h>
```
Is this include needed here?

pfalcon replied:
> Is this include needed here?

A simple generic answer is that this header was here before. [And thus it would be stretching this patch too much to remove it.]

But yes, it's needed to declare stuff like `iotc_mqtt_qos_t` as referenced in this file.

atigyi commented on samples/net/google_iot_device/src/example_utils.c:
```diff
@@ -17,90 +17,6 @@ extern iotc_crypto_key_data_t iotc_connect_private_key_data;
-
-  return 0;
-}
-
```
👍

atigyi commented on samples/net/google_iot_device/src/config.c.in:
```diff
@@ -0,0 +1,19 @@
+
+const char* iotc_publish_message = "Hello from Zephyr!";
+
+iotc_mqtt_qos_t iotc_example_qos = IOTC_MQTT_QOS_AT_LEAST_ONCE;
```
👍

atigyi commented on samples/net/google_iot_device/src/main.c:
```diff
@@ -19,32 +18,8 @@ char ec_private_key_pem[PRIVATE_KEY_BUFFER_SIZE] = {0};
-  }
-
-  if (0 != load_ec_private_key_pem_from_posix_fs(ec_private_key_pem,
-                                                 PRIVATE_KEY_BUFFER_SIZE)) {
```
👍

atigyi reviewed (commented):
Tried to compile and link this branch. Compilation succeeded. Link errors appeared: unresolved external symbols for the config variables:
```
[ 80%] Built target ..__modules__crypto__mbedtls
[ 98%] Built target ..__modules__lib__google__iot-device-sdk-embedded-c__zephyr
[100%] Linking C executable zephyr.elf
/usr/bin/ld: ../app/libapp.a(main.c.obj): in function `strncpy':
/usr/include/bits/string_fortified.h:106: undefined reference to `iotc_private_key_pem'
/usr/bin/ld: ../app/libapp.a(main.c.obj): in function `zephyr_app_main':
/usr/local/google/home/atigyi/git/atigyi/west_zephyr_update/zephyr/samples/net/google_iot_device/src/main.c:50: undefined reference to `iotc_project_id'
/usr/bin/ld: /usr/local/google/home/atigyi/git/atigyi/west_zephyr_update/zephyr/samples/net/google_iot_device/src/main.c:60: undefined reference to `iotc_device_path'
```

Do I miss a step where the config.h.in isn't included? Or do I miss a build step? Or do we need to extend the [build steps](https://github.com/atigyi/zephyr/blob/google_iot_device_sdk_integration/samples/net/google_iot_device/README.rst) for this sample application? Not sure, please check this, thanks.

Update: don't we miss the following line in the CMakeLists.txt?
`configure_file(${APPLICATION_SOURCE_DIR}/src/config.c.in ${APPLICATION_SOURCE_DIR}/src/config.c) `

✅ Ok, tested. With the line added above and `config.c.in` filled in properly connection and MQTT Publish to Cloud IoT was successful.

@pfalcon: please double-check the suggested extension above, thx!

pfalcon commented:
@atigyi, Pushed a change based on your last round of comments (specifically, copyright header/include in config.c.in).

atigyi commented:
Please check [this link-error comment](https://github.com/atigyi/zephyr/pull/8#pullrequestreview-282257414) too.

pfalcon commented:
> Please check this link-error comment too.

Ah, right.

> Update: don't we miss the following line in the CMakeLists.txt?
configure_file(${APPLICATION_SOURCE_DIR}/src/config.c.in ${APPLICATION_SOURCE_DIR}/src/config.c)

>  Ok, tested. With the line added above and config.c.in filled in properly

Well, I'm not familiar with that CMake command. In my list, what's needed to be done is:

* Before building:
* Copy config.c.in to config.c
* Fill in actual connection config in config.c

Editing config.c.in directly should be discouraged - this file is under version control, so it's possible to inadvertently commit it with other changes, then push to public or otherwise not-enough-restricted repo, then we have credentials leak.

So you're right that README should be updated, working on that.

pfalcon committed:
samples: google_iot_device: README: Update for moving from cmdline to config.c

Signed-off-by: Paul Sokolovsky <paul.sokolovsky@linaro.org>

pfalcon commented:
> So you're right that README should be updated, working on that.

Pushed update as a separate commit, we'll yet need to squash things around later anyway.

atigyi commented:
> > Please check this link-error comment too.
> 
> Ah, right.
> 
> > Update: don't we miss the following line in the CMakeLists.txt?
> > configure_file(${APPLICATION_SOURCE_DIR}/src/config.c.in ${APPLICATION_SOURCE_DIR}/src/config.c)
> 
> > Ok, tested. With the line added above and config.c.in filled in properly
> 
> Well, I'm not familiar with that CMake command. In my list, what's needed to be done is:
> 
> * Before building:
> * Copy config.c.in to config.c
> * Fill in actual connection config in config.c
> 
> Editing config.c.in directly should be discouraged - this file is under version control, so it's possible to inadvertently commit it with other changes, then push to public or otherwise not-enough-restricted repo, then we have credentials leak.
> 
> So you're right that README should be updated, working on that.

First of all - and after README update - this is now clear and a good solution and I can accept it. 👍 

But the `.c.in` suffix mislead me. `.h.in` / `.c.in` file extensions are occupied by the [configure_file cmake functionality](https://cmake.org/cmake/help/latest/command/configure_file.html). First sentence of the description says: "Copy a file to another location and modify its contents." Actually this is the two bullet points mentioned above, but in an automated fashion (which isn't necessarily better).

And about the accidental credentials leak: a new file `config.c` or a modified existing file `config.c.in` both are a single step from git-staging (`git add` or `git stage`) and then git-commit. Maybe `git add` is used less frequently. But I don't see significantly larger thread against credentials leak with the manual steps.

Please just consider if this solution is better with auto-copy and fill or manual. I searched for it, no Zephyr sample uses cmake's configure_file so far but internals do.

pfalcon commented:
> But the .c.in suffix mislead me. .h.in / .c.in file extensions are occupied by the configure_file cmake functionality.

Ah, I see. And not exactly. ".in" suffix is a well-known convention for "template" files where things needs to be replaced to get the actual file sans ".in". Where possible, replacement happens automatically. Where not, e.g. where user credentials are needed, it needs to be copied and edited by a user. It's universal convention not tied to any tool. CMake just appear to follow well-known naming convention with its own adhoc meaning. Github gives 158 million hits when searching for "config.h.in": https://github.com/search?q=config.h.in&type=Code , and majority of those hits aren't related to CMake.

> are a single step from git-staging (git add or git stage) and then git-commit. Maybe git add is used less frequently.

Well, those are 2 steps ;-). "git add" is used much less frequently. How many people process git commits is "git commit -a", which just shoves anything changed into single commit, voila. Running "git add &;lt;file>" is quite a departure from that simple course.

> Please just consider if this solution is better with auto-copy and fill or manual. I searched for it, no Zephyr sample uses cmake's configure_file so far but internals do.

I think it's good idea to follow what other samples do and not rely on CMake peculiarities. CMake is quite a recent addition to Zephyr build system. And if it was changed once, it can be changed again, and then CMake-specific features will complicate migration and maintenance.

Also mind that the purpose of this patch is to get rid of the code not supported by target other than `native_posix`, and move on to other problem precluding it to run on other targets, and not perfectalize configuration handling. It's quite possible that when we get stuff actually working, we'll think: "Oh, we want those params to be set via Kconfig". Or maybe we'll be asked to do that during upstream code review...

atigyi reviewed (approved):
👍 Great job, thank you for removing the command line args!!
